docs/writing-fusing-functions.md

```diff
@@ -67,9 +67,58 @@ main = λxλt1λf1 (λtλf (x f t) f1 t1)
 main = λxλt1λf1 (λf (x f f1) t1))
 main = λxλt1λf1 (x t1 f1)
 ```
-Wow! Simply by replacing lambda arguments with the values applied to them, we were able to make `(not (not x))` not grow in size. This is what fusing means, and it's a really powerful tool to make programs faster.
+Wow! Simply by replacing lambda arguments with the values applied to them, we were able to make `(not (not x))` not grow in size. This is what fusing means, and it's a really powerful tool to make programs faster. 
 
-Fusing isn't only for Church-encoded `not`. Fusing can be done anywhere where efficient composition is important. Broadly speaking, a good rule of thumb in HVM is "push lambdas to the top and duplications to the bottom".
+Fusing isn't only for Church-encoded `not`. Fusing can be done anywhere where efficient composition is important. What we simply have to do is "lift" lambdas up, and make the arguments "shared" between all cases.
+
+## Preventing duplication
+
+Something else that is important when writing fusing functions is linearizing variables to prevent useless duplication.
+
+Consider the [scott-encoded](https://crypto.stanford.edu/~blynn/compiler/scott.html) type `Nat`:
+
+```rs
+zero = λs λz z
+succ = λpred λs λz (s pred) // Creates a Scott number out of its predecessor
+
+two = (succ (succ zero)) // λs λz (s λs λz (s λs λz z))
+```
+
+We can write an addition function for this type, that adds two numbers
+
+```rs
+add = λa λb
+	let case_succ = λa_pred (succ (add a_pred b)) // If a = p+1, then return (p+b)+1
+	let case_zero = b // If the `a = 0`, return `b`
+	(a case_succ case_zero)
+```
+
+This code seems fine, but it has a problem that makes it suboptimal. There is an implicit [duplication](dups-and-sups.md) here:
+
+```rs
+add = λa λb
+	let {b1 b2} = b;
+	let case_succ = λa_pred (succ (add a_pred b0))
+	let case_zero = b1
+	(a case_succ case_zero)
+```
+
+However, only one of the two duplicated values is ever used. The other one is discarded. This means that this duplication is completely useless! It makes this function suboptimal. In fact, as of commit `ebd469`, it will crash HVM, because it's recursive and HVM2 is eager.
```

Comment:
It won't crash, it will get stuck in an infinite loop (maybe eventually crash with OoM but that will take a long time)

Reply:
Replaced "crash" by "hang" in 26f4e65